Skaffold API docs (#3068)
GoogleContainerTools/skaffold: docs/content/en/docs/concepts/api/_index.md

---
title: "Skaffold API"
linkTitle: "Skaffold API"
weight: 40
---
When running [`skaffold dev`]({{< relref "/docs/workflows/dev" >}}) or [`skaffold debug`]({{< relref "/docs/workflows/debug" >}}), 
Skaffold starts a server that exposes an API over the lifetime of the Skaffold process.
Besides the CLI, this API is the primary way tools like IDEs integrate with Skaffold for **retrieving information about the
pipeline** and for **controlling the phases in the pipeline**.

To retrieve information of the Skaffold pipeline, the Skaffold API provides two main functionalities:
  

>   * A [streaming event log]({{< relref "/docs/concepts/api#events-api">}}) created from the different phases in a pipeline run and

Author: there is usually no comma before and right?

>   * A [streaming event log]({{< relref "/docs/concepts/api#events-api">}}) created from the different phases in a pipeline run and

Reviewer: in this case since these are numbered items in a list, i think there should be one there. its nitpicky though

  
  * A snapshot of the [overall state]({{< relref "/docs/concepts/api#state-api" >}}) of the pipeline at any given time during the run.

To control the individual phases of the Skaffold, the Skaffold API provides [fine grained control over]({{< relref "/docs/concepts/api#controlling-build-sync-deploy" >}})

> the individual phases of the pipeline (build, deploy and sync).

Author: i will keep this as build, deploy and sync.

{{</alert>}}

```golang
import (
  "log"
  pb "github.com/GoogleContainerTools/skaffold/proto"
  "google.golang.org/grpc"
)

func main(){
  conn, err := grpc.Dial("localhost:50051", grpc.WithInsecure())
  if err != nil {
    log.Fatalf("fail to dial: %v", err)
  }
  defer conn.Close()
  client := pb.NewSkaffoldServiceClient(conn)
}
```

## API structure

Skaffold's API exposes the three main endpoints:

* Events API - stream of lifecycle events
* State API - retrieve the current state
* Control API - control build/deploy/sync

### Events API

Skaffold provides a continuous development mode [`skaffold dev`]({{< relref "/docs/workflows/dev" >}}) which builds, deploys
your application on changes. In a single development loop, one or more container images
may be built and deployed.

Skaffold exposes events for clients to get notified when phases within a development loop
start, succeed or fail.
Tools that integrate with Skaffold can use these events to kick-off parts of the development workflow depending on them.

Example scenarios:

* port forwarding events are used by Cloud Code to attach debuggers automatically to running containers.     

> * when a port-forwarded frontend service is redeployed successfully, kick-off a suite of Selenium tests can be executed to test changes.

Author: done

**Event API contract**

| protocol | endpoint | encoding |
| ---- | --- | --- |
| HTTP | `http://localhost:{HTTP_RPC_PORT}/v1/events` | newline separated JSON using chunk transfer encoding over HTTP|  
| gRPC | `client.Events(ctx)` method on the [`SkaffoldService`]({{< relref "/docs/references/api#skaffoldservice">}}) | protobuf 3 over HTTP |  

**Examples**

{{% tabs %}}
{{% tab "HTTP API" %}}
Using `curl` and `HTTP_RPC_PORT=50052`, an example output of a `skaffold dev` execution on our [getting-started example](https://github.com/GoogleContainerTools/skaffold/tree/master/examples/getting-started)
```bash
 curl localhost:50052/v1/events

> {"result":{"timestamp":"2019-10-16T18:26:11.385251549Z","event":{"metaEvent":{"entry":"Starting Skaffold: \u0026{Version:v0.39.0-16-g5bb7c9e0 ConfigVersion:skaffold/v1beta15 GitVersion: GitCommit:5bb7c9e078e4d522a5ffc42a2f1274fd17d75902 GitTreeState:dirty BuildDate:2019-10-03T15:01:29Z GoVersion:go1.13rc1 Compiler:gc Platform:linux/amd64}"}}}}

Author: done thanks!

---
title: "Skaffold API"
linkTitle: "Skaffold API"
weight: 40
---

When running `skaffold dev` or `skaffold debug`, Skaffold starts a server that exposes an API over the lifetime of the run.

> This API is the primary way tools like IDEs integrate with skaffold, both to subscribe to an event log 

Reviewer 2: ```suggestion
Besides the CLI, this API is the primary way tools like IDEs integrate with Skaffold for **retrieving information about the pipeline** and for **controlling the phases in the pipeline**. To  retrieve information of the Skaffold pipeline, the Skaffold API provides two main functionalities: 
- a streaming event log  
```

> created from the different phases in a pipeline run, and to get a snapshot of the overall state of the pipeline at any given time during the run.

Reviewer 2: ```suggestion
created from the different phases in a pipeline run
- and a snapshot of the overall state of the pipeline at any given time during the run.
```

> The API also provides fine grain controls over the individual components of the Skaffold

Reviewer 2: ```suggestion
The API also provides fine grained control over the individual phases of the Skaffold
```

> pipeline (build, deploy and sync), as opposed to relying on Skaffold’s built-in trigger mechanisms.

Reviewer 2: ```suggestion
pipeline (build, deploy and sync), as opposed to relying on Skaffold’s automation.
```

## Skaffold API 

> Skaffold API is restful and `gRPC` based, so it works with any language that has an HTTP library, such as cURL and urllib.

Reviewer: ```suggestion
The Skaffold API is both restful and `gRPC` based, so it can be accessed directly through a terminal via `curl`, or through your favorite client library for issuing HTTP requests or writing `gRPC` clients.
```

> The API server runs on localhost at predefined ports.

Reviewer: ```suggestion
The server is hosted locally on the Skaffold host machine, and will serve by default on ports 50051 and 50052, though these ports can be configured through the `--rpc-port` and `--rpc-http-port` flags.
```

> The protos used can be found here. (todo add link)

Reviewer: ```suggestion
The server's protocol is based on protobufs: documentation for these can be found here. TODO (tejaldesai): add link
```

### gRPC Server

> gRPC API server is exposed on port `50051` by default. If the port is busy, Skaffold will find the next available port. 

Reviewer: ```suggestion
The gRPC API is exposed on port `50051` by default. If this port is busy, Skaffold will find the next available port. 
```

> You can grab the port from Skaffold logs.

Reviewer: ```suggestion
You can find this port from Skaffold's logs on startup.
```

```code
$ skaffold dev
WARN[0000] port 50051 for gRPC server already in use: using 50053 instead 
``` 

> You can also specify a port on the command line with flag `--rpc-port`.

Reviewer: ```suggestion
You can also specify a port on the command line with the `--rpc-port` flag.
```

> ### gRPC REST Server  

Reviewer: ```suggestion
### HTTP (REST) API  
```

> REST API server is exposed on port `50052` by default. If the port is busy, Skaffold will find the next available port. 

Reviewer: ```suggestion
The HTTP API is exposed on port `50052` by default. As with the gRPC API, if this port is busy, Skaffold will find the next available port, and the final port can be found from Skaffold's startup logs.
```

You can grab the port from Skaffold logs.

```code
$ skaffold dev
WARN[0000] port 50052 for gRPC HTTP server already in use: using 50055 instead 
``` 

> You can also specify a port on the command line with flag `--rpc-http-port`.

Reviewer: ```suggestion
You can also specify a port on the command line with the `--rpc-http-port` flag.
```

## Skaffold API

> Skaffold API Server exposes following endpoints.

Reviewer: ```suggestion
Skaffold's API exposes the following endpoints:
```

### GET /v1/events

> Skaffold provides a continuous development mode [`skaffold dev`](../modes/#skaffold_dev) which builds, deploys

Reviewer 3: It is not clear enough what we get as response of this call. I suppose it is a "long polling" response similar to the "watch" operation on Kubernetes, for example. Is it the case? If so, is it using chunk transfer encoding, or something else?

Author: We will link the protos soon and hope it will make it more clear.

Reviewer 2: we could add more info about this in here: 

on the gRPC layer you have streaming gRPC of the event stream (protos will explain). 
for the HTTP layer we use https://github.com/grpc-ecosystem/grpc-gateway they translate gRPC streams to newline separated JSON stream which is using chunk transfer encoding. See an example curl output with `--rpc-http-port 5555`: 

curl -v http://localhost:5555/v1/events
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 5555 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 5555 (#0)
> GET /v1/events HTTP/1.1
> Host: localhost:5555
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Date: Wed, 16 Oct 2019 20:45:37 GMT
< Transfer-Encoding: chunked
< 
{"result":{"timestamp":"2019-10-16T20:45:15.402079Z","event":{"metaEvent":{"entry":"Starting Skaffold: \u0026{Version:v0.39.0-131-g1759410a7-dirty ConfigVersion:skaffold/v1beta16 GitVersion: GitCommit:1759410a7071b1b7858a2a8b85f8a8bfb07212da GitTreeState:dirty BuildDate:2019-10-10T11:44:31Z GoVersion:go1.13.1 Compiler:gc Platform:darwin/amd64}"}}}}
{"result":{"timestamp":"2019-10-16T20:45:15.847029Z","event":{"buildEvent":{"artifact":"node-guestbook-backend","status":"In Progress"}},"entry":"Build started for artifact node-guestbook-backend"}}
{"result":{"timestamp":"2019-10-16T20:45:15.847022Z","event":{"buildEvent":{"artifact":"node-guestbook-frontend","status":"In Progress"}},"entry":"Build started for artifact node-guestbook-frontend"}}
{"result":{"timestamp":"2019-10-16T20:45:15.859036Z","event":{"buildEvent":{"artifact":"node-guestbook-frontend","status":"Complete"}},"entry":"Build completed for artifact node-guestbook-frontend"}}
{"result":{"timestamp":"2019-10-16T20:45:15.861777Z","event":{"buildEvent":{"artifact":"node-guestbook-backend","status":"In Progress"}},"entry":"Build started for artifact node-guestbook-backend"}}
{"result":{"timestamp":"2019-10-16T20:45:16.316248Z","event":{"buildEvent":{"artifact":"node-guestbook-backend","status":"Complete"}},"entry":"Build completed for artifact node-guestbook-backend"}}
{"result":{"timestamp":"2019-10-16T20:45:16.326313Z","event":{"deployEvent":{"status":"In Progress"}},"entry":"Deploy started"}}
{"result":{"timestamp":"2019-10-16T20:45:17.605745Z","event":{"deployEvent":{"status":"Complete"}},"entry":"Deploy complete"}}
```

Reviewer: We should try to be clear about a few things for all of the endpoints here:

* the exact URL, and the HTTP method
* the expected response (both proto and JSON)
* the expected payload, if any (proto and JSON examples would be helpful here too)

Also, for endpoints like this one, implementation details are useful. For example, when using the /v1/events endpoint over HTTP, it's useful to know that it's using opening a long standing connection (aka the connection is open until the client cancels or the server shuts down), and it's using HTTP/2 server push to send events to an open connection ala pub sub.

Author: Regarding response for, i will link in the EventLog proto from the api ref which just got merged.
i think the payload will come we write the docs for `/v1/execute`

Reviewer 2: in an API spec we need to be precise with the response. 
The EventLog proto will only refer to the gRPC doc, not the JSON format, and definitely doesn't have information about how streaming is implemented for HTTP clients. We need to have that here as well.

> ## Skaffold API

Reviewer 2: we are repeating Skaffold API again here as a header

your application on changes. In a single development loop, one or more container images
may be built and deployed. The time taken for the changes to deploy varies.

Skaffold exposes events for users to get notified when phases within a development loop
complete. 
You can use these events to automate next steps in your development workflow. 

> e.g: when making a change to port-forwarded frontend service, reload the 

Reviewer: ```suggestion
For example, when making a change to port-forwarded frontend service, reload the 
```

browser url after the service is deployed and running to test changes.

> Here is way to get events for a `skaffold dev` [getting-started example](https://github.com/GoogleContainerTools/skaffold/tree/master/examples/getting-started)

Reviewer 2: ```suggestion
An example output using curl to get events for a `skaffold dev` execution on our [getting-started example](https://github.com/GoogleContainerTools/skaffold/tree/master/examples/getting-started)
```

---
title: "Skaffold API"
linkTitle: "Skaffold API"
weight: 40
---

When running `skaffold dev` or `skaffold debug`, Skaffold starts a server that exposes an API over the lifetime of the run.
Besides the CLI, this API is the primary way tools like IDEs integrate with Skaffold for **retrieving information about the pipeline** and for **controlling the phases in the pipeline**. To  retrieve information of the Skaffold pipeline, the Skaffold API provides two main functionalities: 
- a streaming event log  
created from the different phases in a pipeline run
- and a snapshot of the overall state of the pipeline at any given time during the run.
The API also provides fine grained control over the individual phases of the Skaffold
pipeline (build, deploy and sync), as opposed to relying on Skaffold’s automation.

## Skaffold API 

> The Skaffold API is `gRPC` based, and it is also exposed via the gRPC gateway as a JSON over HTTP service.

Reviewer 4: exposed as a JSON over HTTP service, via the gRPC gateway

The server is hosted locally on the same host where the skaffold process is running, and will serve by default on ports 50051 and 50052. These ports can be configured through the `--rpc-port` and `--rpc-http-port` flags.
The server's gRPC service definitions and message protos can be found [here](/#). 
{{% todo 1703 %}}

### gRPC Server

The gRPC API is exposed on port `50051` by default. If this port is busy, Skaffold will find the next available port. 
You can find this port from Skaffold's logs on startup.

```code
$ skaffold dev
WARN[0000] port 50051 for gRPC server already in use: using 50053 instead 
``` 
You can also specify a port on the command line with the `--rpc-port` flag.

### HTTP (REST) API  
The HTTP API is exposed on port `50052` by default. As with the gRPC API, if this port is busy, Skaffold will find the next available port, and the final port can be found from Skaffold's startup logs.

> You can grab the port from Skaffold logs.

Reviewer 4: Remove this line?
